apps/hash-api/src/integrations/google/oauth-callback.ts

```diff
@@ -0,0 +1,185 @@
+import { createGoogleOAuth2Client } from "@local/hash-backend-utils/google";
+import { getMachineActorId } from "@local/hash-backend-utils/machine-actors";
+import {
+  GoogleOAuth2CallbackRequest,
+  GoogleOAuth2CallbackResponse,
+} from "@local/hash-isomorphic-utils/google-integration";
+import { googleEntityTypes } from "@local/hash-isomorphic-utils/ontology-type-ids";
+import { AccountProperties as GoogleAccountProperties } from "@local/hash-isomorphic-utils/system-types/googlesheetsintegration";
+import { OwnedById } from "@local/hash-subgraph";
+import { RequestHandler } from "express";
+import { google } from "googleapis";
+
+import { createEntity } from "../../graph/knowledge/primitive/entity";
+import { createUserSecret } from "../../graph/knowledge/system-types/user-secret";
+import { enabledIntegrations } from "../enabled-integrations";
+import { getGoogleAccountById } from "./shared/get-google-account";
+
+export const googleOAuthCallback: RequestHandler<
+  Record<string, never>,
+  GoogleOAuth2CallbackResponse,
+  GoogleOAuth2CallbackRequest
+> =
+  // @todo upgrade to Express 5, which handles errors from async request handlers automatically
+  // eslint-disable-next-line @typescript-eslint/no-misused-promises
+  async (req, res) => {
+    if (!req.user) {
+      res.status(401).send({ error: "User not authenticated." });
+      return;
+    }
+
+    const { vaultClient } = req.context;
+
+    if (!vaultClient) {
+      res.status(501).send({ error: "Vault integration is not configured." });
+      return;
+    }
+
+    if (!enabledIntegrations.googleSheets) {
+      res.status(501).send({ error: "Google integration is not enabled." });
+      return;
+    }
+
+    const { code } = req.body;
+
+    const googleOAuth2Client = createGoogleOAuth2Client();
+
+    const oauth2 = google.oauth2({
+      auth: googleOAuth2Client,
+      version: "v2",
+    });
+
+    const { tokens } = await googleOAuth2Client.getToken({
+      code,
+    });
+
+    if (!tokens.refresh_token || !tokens.access_token) {
+      res.status(500).send({
+        error: "Could not exchange code for access and refresh tokens",
+      });
+      return;
+    }
+
+    googleOAuth2Client.setCredentials(tokens);
+
+    const googleUser = await oauth2.userinfo.get();
+
+    if (
+      !googleUser.data.id ||
+      !googleUser.data.email ||
+      !googleUser.data.name
+    ) {
+      res.status(500).send({
+        error: "Google user data is missing required fields",
+      });
+      return;
+    }
```

Comment:
**Blocking:** the `name` field can be `undefined` for some google accounts (e.g. for my `bw@hash.ai` google account). We could account for this by:
- using the email address as the display name, if no `name` is available from google
- making the `displayName` property optional on the google account entity

Reply:
Thank you – I've gone for defaulting to the email address, in https://github.com/hashintel/hash/pull/4141/commits/384f696db89ce4fe68dd0483470dc30e4670d49a